File: backend/Dia5.Application/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Dia5.Application.DTOs;
using Dia5.Application.Interfaces;
using Dia5.Domain.Entities;
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;

namespace Dia5.Application.Services;

public class AuthService : IAuthService
{
    private readonly IUsuarioRepository _usuarioRepository;
    private readonly IConfiguration _configuration;

    public AuthService(IUsuarioRepository usuarioRepository, IConfiguration configuration)
    {
        _usuarioRepository = usuarioRepository;
        _configuration = configuration;
    }

    public async Task<AuthResponseDto> RegisterAsync(RegisterUserDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Senha))
        {
            throw new InvalidOperationException("E-mail e senha são obrigatórios.");
        }

        var usuarioExistente = await _usuarioRepository.GetByEmailAsync(dto.Email);
        if (usuarioExistente != null)
        {
            throw new InvalidOperationException("Já existe um usuário cadastrado com este e-mail.");
        }

        var senhaHash = BCrypt.Net.BCrypt.HashPassword(dto.Senha);
        var codigoPerfil = await GerarCodigoPerfilUnicoAsync();

        var usuario = new Usuario
        {
            Nome = dto.Nome,
            Email = dto.Email,
            SenhaHash = senhaHash,
            CodigoPerfil = codigoPerfil,
            IsGuest = false
        };

        usuario.Validar();

        await _usuarioRepository.AddAsync(usuario);
        await _usuarioRepository.SaveChangesAsync();

        var token = GerarJwtToken(usuario);

        return new AuthResponseDto
        {
            Token = token,
            UserId = usuario.Id,
            Nome = usuario.Nome,
            Email = usuario.Email,
            CodigoPerfil = usuario.CodigoPerfil
        };
    }

    public async Task<AuthResponseDto> LoginAsync(LoginDto dto)
    {
        var usuario = await _usuarioRepository.GetByEmailAsync(dto.Email);
        if (usuario == null || string.IsNullOrEmpty(usuario.SenhaHash))
        {
            throw new InvalidOperationException("E-mail ou senha inválidos.");
        }

        var senhaValida = BCrypt.Net.BCrypt.Verify(dto.Senha, usuario.SenhaHash);
        if (!senhaValida)
        {
            throw new InvalidOperationException("E-mail ou senha inválidos.");
        }

        var token = GerarJwtToken(usuario);

        return new AuthResponseDto
        {
            Token = token,
            UserId = usuario.Id,
            Nome = usuario.Nome,
            Email = usuario.Email ?? string.Empty,
            CodigoPerfil = usuario.CodigoPerfil ?? string.Empty
        };
    }

    private async Task<string> GerarCodigoPerfilUnicoAsync()
    {
        string codigo;
        do
        {
            codigo = Guid.NewGuid().ToString("N")[..6].ToUpper();
        } while (await _usuarioRepository.GetByCodigoPerfilAsync(codigo) != null);

        return codigo;
    }

    private string GerarJwtToken(Usuario usuario)
    {
        var secretKey = _configuration["Jwt:SecretKey"] ?? "MinhaChaveSecretaSuperSeguraDia5App2026!";
        var key = Encoding.UTF8.GetBytes(secretKey);

        var tokenHandler = new JwtSecurityTokenHandler();
        var tokenDescriptor = new SecurityTokenDescriptor
        {
            Subject = new ClaimsIdentity(new[]
            {
                new Claim(ClaimTypes.NameIdentifier, usuario.Id.ToString()),
                new Claim(ClaimTypes.Name, usuario.Nome),
                new Claim(ClaimTypes.Email, usuario.Email ?? string.Empty)
            }),
            Expires = DateTime.UtcNow.AddDays(7),
            SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
        };

        var token = tokenHandler.CreateToken(tokenDescriptor);
        return tokenHandler.WriteToken(token);
    }
}

File: backend/src/modules/auth/auth.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User } from '../../domain/entities/user.entity';
import { RegisterUserDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';
import { AuthResponseDto } from './dto/auth-response.dto';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcryptjs';
import { randomUUID } from 'crypto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    private readonly jwtService: JwtService,
  ) {}

  async register(dto: RegisterUserDto): Promise<AuthResponseDto> {
    if (!dto.email || !dto.senha) {
      throw new BadRequestException('E-mail e senha são obrigatórios.');
    }

    const existingUser = await this.userRepository.findOne({ where: { email: dto.email } });
    if (existingUser) {
      throw new BadRequestException('Já existe um usuário cadastrado com este e-mail.');
    }

    const salt = await bcrypt.genSalt(10);
    const senhaHash = await bcrypt.hash(dto.senha, salt);
    const codigoPerfil = await this.generateUniqueProfileCode();

    const user = new User();
    user.id = randomUUID();
    user.nome = dto.nome;
    user.email = dto.email;
    user.senhaHash = senhaHash;
    user.codigoPerfil = codigoPerfil;
    user.isGuest = false;

    // Roda validações de regras de domínio
    user.validar();

    await this.userRepository.save(user);

    const token = this.generateJwtToken(user);

    return {
      token,
      userId: user.id,
      nome: user.nome,
      email: user.email!,
      codigoPerfil: user.codigoPerfil!,
    };
  }

  async login(dto: LoginDto): Promise<AuthResponseDto> {
    const user = await this.userRepository.findOne({ where: { email: dto.email } });
    if (!user || !user.senhaHash) {
      throw new BadRequestException('E-mail ou senha inválidos.');
    }

    const isPasswordValid = await bcrypt.compare(dto.senha, user.senhaHash);
    if (!isPasswordValid) {
      throw new BadRequestException('E-mail ou senha inválidos.');
    }

    const token = this.generateJwtToken(user);

    return {
      token,
      userId: user.id,
      nome: user.nome,
      email: user.email || '',
      codigoPerfil: user.codigoPerfil || '',
    };
  }

  private async generateUniqueProfileCode(): Promise<string> {
    let codigo: string;
    let exists = true;
    while (exists) {
      codigo = randomUUID().replace(/-/g, '').substring(0, 6).toUpperCase();
      const existing = await this.userRepository.findOne({ where: { codigoPerfil: codigo } });
      if (!existing) {
        exists = false;
      }
    }
    return codigo!;
  }

  private generateJwtToken(user: User): string {
    const payload = {
      sub: user.id,
      nome: user.nome,
      email: user.email,
    };
    return this.jwtService.sign(payload);
  }
}
